token-lending/program/src/processor.rs

```diff
@@ -1310,175 +1258,170 @@ fn process_withdraw_obligation_collateral(
+    if &obligation_token_input.mint != obligation_token_mint_info.key {
+        msg!("Obligation token input mint does not match the obligation token mint provided");
+        return Err(LendingError::InvalidTokenMint.into());
     }
 
-    let obligation_token_amount = obligation
-        .collateral_to_obligation_token_amount(collateral_amount, obligation_token_mint.supply)?;
-
-    obligation.deposited_collateral_tokens = remaining_collateral;
-
-    Obligation::pack(obligation, &mut obligation_info.data.borrow_mut())?;
-
     let authority_signer_seeds = &[
         lending_market_info.key.as_ref(),
         &[lending_market.bump_seed],
     ];
     let lending_market_authority_pubkey =
         Pubkey::create_program_address(authority_signer_seeds, program_id)?;
-    if lending_market_authority_info.key != &lending_market_authority_pubkey {
+    if &lending_market_authority_pubkey != lending_market_authority_info.key {
+        msg!(
+            "Derived lending market authority does not match the lending market authority provided"
+        );
         return Err(LendingError::InvalidMarketAuthority.into());
     }
 
-    // burn obligation tokens
+    let withdraw_amount = if obligation.borrows.is_empty() {
+        // there are no borrows; they have been repaid, liquidated, or were never taken out
+        collateral.deposited_amount
+    } else if obligation.borrowed_value == Decimal::zero() {
+        // there are borrows, but they cannot be valued; they must be repaid to withdraw collateral
+        msg!("Obligation borrowed value is zero");
+        return Err(LendingError::ObligationBorrowsZero.into());
+    } else if obligation.deposited_value == Decimal::zero() {
+        // there are deposits, but they cannot be valued
+        msg!("Obligation deposited value is zero");
+        return Err(LendingError::ObligationDepositsZero.into());
+    } else {
+        // there are borrows and deposits, and they can both be valued
+        let max_withdraw_value = obligation.max_withdraw_value()?;
+        if max_withdraw_value == Decimal::zero() {
+            msg!("Maximum withdraw value is zero");
+            return Err(LendingError::WithdrawTooLarge.into());
+        }
+
+        let withdraw_amount = if collateral_amount == u64::MAX {
+            let withdraw_value = max_withdraw_value.min(collateral.market_value);
+            let withdraw_pct = withdraw_value.try_div(collateral.market_value)?;
+            withdraw_pct
+                .try_mul(collateral.deposited_amount)?
+                .try_floor_u64()?
+                .min(collateral.deposited_amount)
+        } else {
+            let withdraw_amount = collateral_amount.min(collateral.deposited_amount);
+            let withdraw_pct =
+                Decimal::from(withdraw_amount).try_div(collateral.deposited_amount)?;
+            let withdraw_value = collateral.market_value.try_mul(withdraw_pct)?;
+            if withdraw_value > max_withdraw_value {
```

Comment:
I was wondering about this and I'm okay with making this change, but someone can always withdraw `u64::MAX` to represent that intention.

Is this like borrow and deposit where we fail hard if we can't fulfill the request, or is this like liquidate where we always give you as much as you can get up to the requested amount? The case of liquidation is special IMO because it's harder for a liquidator to reason about how much they can get.

Reply:
Yeah, that's a good point. A user may just want the instruction to fail if there isn't enough withdraw available. It feels like withdraw, borrow, deposit, and liquidate should handle this consistently now that you mention it. I'm now leaning towards failing hard, you raise a good point that they can always do MAX for this kind of behavior.